### DISCUSSION TITLE
Scrapli NETCONF copy-config issue with config as multiline-string

### DISCUSSION THREAD

**Discussion by willikubny** [scrapli/scrapli_netconf#120] [Q&A]
Hi everyone

I'm working with Nornir, but as no copy-config task exist at the moment, I've tried to work directly with Scrapli NETCONF.

I'm trying to copy the whole Cisco-IOS-XE-native NETCONF configuration to the candidate datastore and get the following traceback:
`Traceback (most recent call last):
  File "/Users/willikubny/PythonVirtualEnv/zis_iac/./dev_copy_config.py", line 1642, in <module>
    main()
  File "/Users/willikubny/PythonVirtualEnv/zis_iac/./dev_copy_config.py", line 1624, in main
    result = conn.copy_config(source=SOURCE_CONFIG, target="candidate",)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli_netconf/driver/sync_driver.py", line 369, in copy_config
    raw_response = self.channel.send_input_netconf(response.channel_input)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli_netconf/channel/sync_channel.py", line 238, in send_input_netconf
    buf, _ = super().send_input(channel_input=channel_input, strip_prompt=False, eager=True)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/decorators.py", line 235, in decorate
    return wrapped_func(*args, **kwargs)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/channel/sync_channel.py", line 478, in send_input
    self.write(channel_input=channel_input)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/channel/base_channel.py", line 336, in write
    self.transport.write(channel_input=channel_input.encode())
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli_netconf/transport/plugins/system/transport.py", line 67, in write
    self.session.write(bytes_to_send.read(self.write_chunk_size))
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/transport/plugins/system/ptyprocess.py", line 517, in write
    n = self.fileobj.write(bytes_to_write)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/decorators.py", line 75, in _signal_raise_exception
    return _handle_timeout(transport=transport, logger=logger, message=message)
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/decorators.py", line 133, in _handle_timeout
    transport.close()
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/transport/plugins/system/transport.py", line 140, in close
    self.session.close()
  File "/Users/willikubny/.envs/zis_iac/lib/python3.9/site-packages/scrapli/transport/plugins/system/ptyprocess.py", line 439, in close
    raise PtyProcessError("Could not terminate the child.")
scrapli.transport.plugins.system.ptyprocess.PtyProcessError: Could not terminate the child.`

If I just copy the running datastore to the candidate datastore with copy-config, then it's working, which makes me guessing that in general it's working and I have some sort of issue with my source of the copy-config.

My test script is simple:

```python
#!/usr/bin/env python3

from scrapli_netconf import NetconfDriver

IOSXE_DEVICE = {
    "host": "10.1.5.199",
    "auth_username": "nornir",
    "auth_password": "bwd8woua",
    "auth_strict_key": False,
}

SOURCE_CONFIG = """
<config>
  <native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native">
    ### Whole config ###
  </native>
</config>
"""

def main():
    """Basic use example"""
    # create scrapli_netconf connection just like with scrapli, open the connection
    conn = NetconfDriver(**IOSXE_DEVICE)
    conn.open()

    """
    # lock the candidate config before starting because why not
    result = conn.lock(target="candidate")
    print(result.result)
    """

    result = conn.copy_config(source=SOURCE_CONFIG, target="candidate",)
    print(result.result)

    # commit config changes
    result = conn.commit()
    print(result.result)

    """
    # unlock the candidate now that we're done
    result = conn.unlock(target="candidate")
    print(result.result)
    """

    # close the session
    conn.close()


if __name__ == "__main__":
    main()
```

I didn't found any helpful information yet and appreciate every feedback.
An example of what I try to do would be amazing.

**Comment by carlmontanari** [comment]
Probably want logs to see whats up, example -> [here](https://github.com/carlmontanari/scrapli/blob/main/examples/logging/opinionated_logging.py)

It looks like things time out during the copy config operation so yeah, just gotta see whats up with that. Could also try to crank the timeout up if its a big config or the device is being a bit slow about it.

HTH

**Reply by willikubny** [reply]
Hi Carl

Thanks for you fast answer. I've added the logging and I've also added the transport ssh2. With the transport ssh2 I don't get anymore a timeout, but a a rpc-error reply.

`<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="101">
  <rpc-error>
    <error-type>application</error-type>
    <error-tag>operation-failed</error-tag>
    <error-severity>error</error-severity>
    <error-path>
    /rpc/copy-config/source
  </error-path>
    <error-message xml:lang="en">Unexpected content: '/&gt;'</error-message>
    <error-info>
      <bad-element>source</bad-element>
    </error-info>
  </rpc-error>
</rpc-reply>`

I could not find out why /&gt; is added at the end to my config payload.
Please find attached the log file.

[scrapli.log](https://github.com/scrapli/scrapli_netconf/files/9336731/scrapli.log)

Do you plan to write a Nornir task also for copy-config?

Thank you in advance.

**Reply by carlmontanari** [reply]
> With the transport ssh2 I don't get anymore a timeout

Cool, thats good news. Probably cranking up the timeout would be fine for system/paramiko transports as well then.


> I could not find out why /> is added at the end to my config payload

Good question, definitely seems to be the culprit. I'm fairly certain I've really only ever used this to copy from datastore to datastore, so we may need to do some investigation around whats going on here. I don't see anything immediately obvious in the code, probably your best is to drop a breakpoint in there and see whats what. If you can provide the absolute minimum reproducible example I may be able to take a peak more this weekend.


> Do you plan to write a Nornir task also for copy-config?

I do not, but happy to review a PR if you're up for it.
